### PR TITLE
Add TimeDependence to RotatedIntervals

### PR DESCRIPTION
**Pull request by kidder**
## Proposed changes

- Add TimeDependence to RotatedIntervals
- Fix a clang-tidy error in src/Elliptic/DiscontinuousGalerkin/DgOperator.hpp 

### Upgrade instructions

If using a RotatedIntervals DomainCreator, you now need to add a TimeDependence option with a value of None to reproduce previous behavior.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/Elliptic/DiscontinuousGalerkin/DgOperator.hpp [hunk 1]
```diff
@@ -469,7 +469,7 @@ struct DgOperatorImpl<System, Linearized, tmpl::list<PrimalFields...>,
                 Spectral::needs_projection(face_mesh, mortar_mesh, mortar_size)
                     ? boundary_data.project_to_mortar(face_mesh, mortar_mesh,
                                                       mortar_size)
-                    : std::move(boundary_data);
+                    : boundary_data;
```
Comment (nilsvu) [inline src/Elliptic/DiscontinuousGalerkin/DgOperator.hpp]: I think the `std::move` is fine and you can `NOLINT` it, perhaps with the following comment:

When no projection is necessary we can safely move the boundary data from the face, because we only have a single neighbor in this direction.

**Review by nilsvu — DISMISSED** [review]

**Comment by nilsvu** [comment]
@kidder clang-tidy is still not appeased. I'm pretty sure the code path is safe, or can you spot a loop-hole? Perhaps it just needs another `NOLINTNEXTLINE`.

@knelli2 Could you provide the second review for this PR? If you haven't done a review yet, here's a few resources:

- [GitHub's guide on PR reviews](https://docs.github.com/en/github/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/reviewing-proposed-changes-in-a-pull-request)
- Our [code review guide](https://spectre-code.org/code_review_guide.html)
- Our [guidelines on the review process](https://spectre-code.org/contributing_to_spectre.html#pull-request-reviews)

Also, feel free to use reviews to ask questions on code you find unclear :)

File: src/Domain/Creators/RotatedIntervals.hpp [hunk 2]
```diff
@@ -129,7 +139,9 @@ class RotatedIntervals : public DomainCreator<1> {
       typename UpperBound::type upper_x,
       typename InitialRefinement::type initial_refinement_level_x,
       typename InitialGridPoints::type initial_number_of_grid_points_in_x,
-      typename IsPeriodicIn::type is_periodic_in) noexcept;
+      typename IsPeriodicIn::type is_periodic_in,
+      std::unique_ptr<domain::creators::time_dependence::TimeDependence<1>>
+          time_dependence) noexcept;
```
Comment (knelli2) [inline src/Domain/Creators/RotatedIntervals.hpp]: Is there a reason to declare the `std::unique_ptr`s in the constructors different from the other arguments? E.g. using `typename TimeDependence::type time_dependence`

File: src/Domain/Creators/RotatedIntervals.hpp [hunk 3]
```diff
@@ -140,6 +152,8 @@ class RotatedIntervals : public DomainCreator<1> {
           lower_boundary_condition,
       std::unique_ptr<domain::BoundaryConditions::BoundaryCondition>
           upper_boundary_condition,
+      std::unique_ptr<domain::creators::time_dependence::TimeDependence<1>>
+          time_dependence,
```
Comment (knelli2) [inline src/Domain/Creators/RotatedIntervals.hpp:156]: Same comment as above.

File: src/Domain/Creators/RotatedIntervals.hpp [hunk 4]
```diff
@@ -172,6 +190,8 @@ class RotatedIntervals : public DomainCreator<1> {
       lower_boundary_condition_{nullptr};
   std::unique_ptr<domain::BoundaryConditions::BoundaryCondition>
       upper_boundary_condition_{nullptr};
+  std::unique_ptr<domain::creators::time_dependence::TimeDependence<1>>
+      time_dependence_;
```
Comment (knelli2) [inline src/Domain/Creators/RotatedIntervals.hpp]: Default to `nullptr`? (which corresponds to `None` time dependence)

File: tests/Unit/Domain/Creators/Test_RotatedIntervals.cpp [hunk 5]
```diff
@@ -58,17 +69,27 @@ void test_rotated_intervals_construction(
   test_domain_construction(
       domain, expected_block_neighbors, expected_external_boundaries,
       make_vector(
-          make_coordinate_map_base<Frame::Logical, Frame::Inertial>(
+          make_coordinate_map_base<
+              Frame::Logical,
+              tmpl::conditional_t<sizeof...(FuncsOfTime) == 0, Frame::Inertial,
+                                  Frame::Grid>>(
```
Comment (knelli2) [inline tests/Unit/Domain/Creators/Test_RotatedIntervals.cpp:75]: Why does having no functions of time indicate that the target frame is the inertial frame, but having some means the target frame is the grid frame? That seems counter-intuitive to me.

File: tests/Unit/Domain/Creators/Test_RotatedIntervals.cpp [hunk 6]
```diff
@@ -58,17 +69,27 @@ void test_rotated_intervals_construction(
...
               CoordinateMaps::DiscreteRotation<1>{OrientationMap<1>{
                   std::array<Direction<1>, 1>{{Direction<1>::lower_xi()}}}},
               CoordinateMaps::Affine{-1., 1., midpoint[0], upper_bound[0]})),
-      0.0, {}, {}, expected_boundary_conditions);
+      10.0, rotated_intervals.functions_of_time(),
```
Comment (knelli2) [inline tests/Unit/Domain/Creators/Test_RotatedIntervals.cpp:84]: Would having the time be at the expiration time of the function of time potentially cause any issues? Maybe pick a time inside the domain of validity?

File: tests/Unit/Domain/Creators/Test_RotatedIntervals.cpp [hunk 7]
```diff
@@ -258,6 +283,12 @@ void test_rotated_intervals_factory() {
         "  UpperBound: [1.0]\n"
         "  InitialGridPoints: [[3,2]]\n"
         "  InitialRefinement: [2]\n"
+        "  TimeDependence:\n"
+        "    UniformTranslation:\n"
+        "      InitialTime: 1.0\n"
+        "      InitialExpirationDeltaT: 9.0\n"
+        "      Velocity: [2.3]\n"
+        "      FunctionOfTimeNames: [TranslationX]\n"
```
Comment (knelli2) [inline tests/Unit/Domain/Creators/Test_RotatedIntervals.cpp:377]: Change the INFO of this test to reflect that it does have time dependence.

File: tests/Unit/Domain/Creators/Test_RotatedIntervals.cpp [hunk 8]
```diff
@@ -232,7 +255,8 @@ void test_rotated_intervals_factory() {
         "  UpperBound: [1.0]\n"
         "  IsPeriodicIn: [True]\n"
         "  InitialGridPoints: [[3,2]]\n"
-        "  InitialRefinement: [2]\n");
+        "  InitialRefinement: [2]\n"
+        "  TimeDependence: None\n");
```
Comment (knelli2) [inline tests/Unit/Domain/Creators/Test_RotatedIntervals.cpp:271]: Change the INFO of this test to reflect that it has no time dependence.

File: tests/Unit/Domain/Creators/Test_RotatedIntervals.cpp [hunk 9]
```diff
@@ -293,6 +324,14 @@ void test_rotated_intervals_factory() {
...
+            std::pair<std::string,
+                      domain::FunctionsOfTime::PiecewisePolynomial<2>>{
+                "TranslationX",
+                {1.0, std::array<DataVector, 3>{{{0.0}, {2.3}, {0.0}}}, 10.0}}),
+        make_vector_coordinate_map_base<Frame::Grid, Frame::Inertial>(
+            CoordinateMaps::TimeDependent::Translation{"TranslationX"},
+            CoordinateMaps::TimeDependent::Translation{"TranslationX"}),
         expected_boundary_conditions);
   }
 }
```
Comment (knelli2) [inline tests/Unit/Domain/Creators/Test_RotatedIntervals.cpp:406]: You have factory tests for:
No boundary condition/no time dependence
With boundary condition/with time dependence

Do you need to test for the other two combinations? (no boundary condition/with time dependence and with boundary condition/no time dependence)

**Review by knelli2 — CHANGES_REQUESTED** [review]
Looks good. Just a couple comments and question for my own understanding.

File: src/Domain/Creators/RotatedIntervals.hpp (hunk 2 above)
Comment (nilsvu) [reply src/Domain/Creators/RotatedIntervals.hpp]: We have been moving away from the Tag::type pattern recently in favor of writing the types explicitly in the constructor. It makes it clearer which types are trivially copyable, which need to get moved, etc. Only when the type is particularly complicated do we keep the Tag::type, e.g. when it is a std::variant.

Comment (kidder) [reply src/Domain/Creators/RotatedIntervals.hpp]: I've remove all the `Tag::type` in favor of the actual type so the file is consistent

File: tests/Unit/Domain/Creators/Test_RotatedIntervals.cpp (hunk 5 above)
Comment (kidder) [reply tests/Unit/Domain/Creators/Test_RotatedIntervals.cpp:75]: The Blocks in a Domain either hold a Logical -> Inertial map if there are no functions of time, or two maps Logical -> Grid and Grid -> Inertial if there are functions of time.   `test_domain_construction` checks for the correct map(s) based on the frame of the passed in map.

File: tests/Unit/Domain/Creators/Test_RotatedIntervals.cpp (hunk 6 above)
Comment (kidder) [reply tests/Unit/Domain/Creators/Test_RotatedIntervals.cpp:84]: The expiration time is considered to be inside the domain of validity.

**Comment by kidder** [comment]
@knelli2 thank you for the excellent review; I have pushed a fixup commit

**Review by knelli2 — DISMISSED** [review]
LGTM :) you can squash. Thanks for the answers to my questions!

**Review by nilsvu — APPROVED** [review]
